### PR TITLE
NextJS 13 App router / RSC support

### PR DESCRIPTION
**Pull request by SutuSebastian**
- [x] I have followed the [Your First Code Contribution section of the Contributing guide](https://github.com/themesberg/flowbite-react/blob/main/CONTRIBUTING.md#your-first-code-contribution)

### Target
NextJS 13 App router and/or RSC support.

### Fixes issues
- #448

### Discussions 
- #923

### Breaking changes (for maintainers)
- removal of `useTheme` react context provider for passing down `theme` to components
- add `'use client';` in all UI components that are stateful and/or have event handlers

### New additions
- consistent theme mode (light/dark/auto) in LS and across the pages
- theme mode (light/dark/auto) sync between multiple browser tabs (same domain)

### Fixes / Improvements
- simplify import/export module hierarchy
- simplify package dependency graph, avoid circular imports errors
- improve iterators performance
- add tests for: `cloneDeep`, `isObject`

### Show off
- simplify DX onboarding, there is no need for `<Flowbite>` component to be rendered in order for the library components to work
  - import only when global theme override is wanted (can also be nested)
- `useThemeMode` hook is context free
  - but acts like it uses one

### CODE REVIEW

**Comment by rluders** [comment]
This is going to be huge. o_O

**Comment by SutuSebastian** [comment]
> This is going to be huge. o_O

Most definitely, but hopefully no changes when it comes to the end user.

**Comment by zoltanszogyenyi** [comment]
> > This is going to be huge. o_O
> 
> Most definitely, but hopefully no changes when it comes to the end user.

That'd be the best outcome - no breaking changes other than documenting a new way of support SSC. Way to go!

**Comment by SutuSebastian** [comment]
> @SutuSebastian just one tip here, don't use merge to get the changes from the main branch, use rebase. Otherwise, it will mess up with the commit history.

Good to know!

**Comment by rluders** [comment]
@SutuSebastian you will need to rebase to remove the merges. Try to avoid merging, add the original flowbite-react repository like this as a second origin to your fork:

```sh
git remote add upstream https://github.com/themesberg/flowbite-react.git
```

Then you can get the changes from the original main to your branch just using the command:

```sh
git pull upstream main --rebase
```

**Comment by SutuSebastian** [comment]
> @SutuSebastian you will need to rebase to remove the merges. Try to avoid merging, add the original flowbite-react repository like this as a second origin to your fork:
> 
> ```shell
> git remote add upstream https://github.com/themesberg/flowbite-react.git
> ```
> 
> Then you can get the changes from the original main to your branch just using the command:
> 
> ```shell
> git pull upstream main --rebase
> ```

I had to solve some conflicts, but from this point on I'll rebase. Thanks for the help!

**Comment by rluders** [comment]
Also, you will probably have to remove the merge commits from your current PR, it will be a little bit painful. You probably will have to do some interactive rebase. You will need to remove/rebase:

[ef50757](https://github.com/themesberg/flowbite-react/pull/994/commits/ef50757540721dada216c1d2214de1611e155fdb)
[011b064](https://github.com/themesberg/flowbite-react/pull/994/commits/011b064b4e53a4632c8d7f34374e6cd1d24e5f10)
[0cfca32](https://github.com/themesberg/flowbite-react/pull/994/commits/0cfca32075fa84abea6609f7fddb0d896d03a917)

File: src/components/FloatingLabel/index.ts
Comment (rluders) [inline src/components/FloatingLabel/index.ts:1]: This is related to #1028

File: src/components/Flowbite/FlowbiteTheme.ts [hunk 1]
```diff
@@ -1,44 +1,41 @@
...
+import type { FlowbiteBreadcrumbTheme } from '../Breadcrumb';
+import type { FlowbiteButtonGroupTheme, FlowbiteButtonTheme } from '../Button';
+import type { FlowbiteCardTheme } from '../Card';
+import type { FlowbiteCarouselTheme } from '../Carousel';
+import type { FlowbiteCheckboxTheme } from '../Checkbox';
+import type { FlowbiteDarkThemeToggleTheme } from '../DarkThemeToggle';
+import type { FlowbiteDatepickerTheme } from '../Datepicker';
+import type { FlowbiteDropdownTheme } from '../Dropdown';
+import type { FlowbiteFileInputTheme } from '../FileInput';
+import type { FlowbiteFloatingLabelTheme } from '../FloatingLabel';
```
Comment (rluders) [inline src/components/Flowbite/FlowbiteTheme.ts:15]: Also #1028

File: src/components/FloatingLabel/index.ts
Comment (rluders) [reply src/components/FloatingLabel/index.ts:1]: BTW, not saying to remove, it is just to keep track. :)

Comment (SutuSebastian) [reply src/components/FloatingLabel/index.ts:1]: Nice, I was actually wondering where are the ticket issues for those

**Comment by rluders** [comment]
@SutuSebastian I tested the branch with all my "huge" experience with NextJS and also with VIte. It seems to be working as expected. Of course, I didn't try all the components, and as we already discussed the fact that some/most of the components still use the `use client` directive probably will not make them work as SC. 

My main test was to see if there wasn't any broken related to the theme. As far as I noticed the only missing feature with the theme support now is the light/dark theme switching. Am I right? So, I think that, if we solve this we can get this PR merged and start to work/test the idea to remove the `use client` directive from the components that should be able to be used as SC.

But this is huge, and very nice work. So, let's finish it and get it merged, then we can work on some other minor features to introduce the SC rendering to the library.

**Comment by rluders** [comment]
@tulup-conner @nigellima I would like to get some feedback from you folks.

**Comment by SutuSebastian** [comment]
> @SutuSebastian I tested the branch with all my "huge" experience with NextJS and also with VIte. It seems to be working as expected. Of course, I didn't try all the components, and as we already discussed the fact that some/most of the components still use the `use client` directive probably will not make them work as SC.
> 
> My main test was to see if there wasn't any broken related to the theme. As far as I noticed the only missing feature with the theme support now is the light/dark theme switching. Am I right? So, I think that, if we solve this we can get this PR merged and start to work/test the idea to remove the `use client` directive from the components that should be able to be used as SC.
> 
> But this is huge, and very nice work. So, let's finish it and get it merged, then we can work on some other minor features to introduce the SC rendering to the library.

Sounds good!

Btw, we do have the theme switching, the hooks is called `useThemeMode`.

I also need to solve some conflicts as well, and "maybe" get the courage to rebase this branch without messing it up :D

**Comment by nigellima** [comment]
> @tulup-conner @nigellima I would like to get some feedback from you folks.

Unfortunately my experience with Next is still minimum. I wish I could have a more solid experience so I could give good feedback here

**Review by rluders — CHANGES_REQUESTED** [review]

File: src/helpers/use-watch-LS-value.ts
[the remark this answers is not in the document]
Comment (SutuSebastian) [reply src/helpers/use-watch-LS-value.ts:1]: It is specific to watching a single value, that's why it's called like that, but I can rename it to `use-watch-localstorage-value.ts`, maybe it makes more sense 🤔

[the remark this answers is not in the document]
Comment (rluders) [reply src/helpers/use-watch-LS-value.ts:1]: Please. :)

**Review by rluders — APPROVED** [review]

**Comment by rluders** [comment]
For all those who are looking for NextJS RSC support, please, check our beta releases:
https://www.npmjs.com/package/flowbite-react/v/0.7.0-beta.3

**Pull request merged**